### PR TITLE
Add gmail to salesforce lead sample

### PR DESCRIPTION
**Pull request by kaneeldias**
## Purpose
$subject

### CODE REVIEW

File: salesforce_api/gmail-to-salesforce-lead/main.bal [hunk 1]
```diff
@@ -0,0 +1,192 @@
+import ballerina/log;
+import ballerinax/googleapis.gmail as gmail;
+import ballerina/mime;
+import ballerinax/salesforce as sfdc;
+import ballerina/lang.runtime;
+import ballerinax/openai.chat as openAI;
```
Comment (anupama-pathirage) [inline salesforce_api/gmail-to-salesforce-lead/main.bal]: Order the imports alphabetically.

File: salesforce_api/gmail-to-salesforce-lead/Config.toml.example [hunk 2]
```diff
@@ -0,0 +1,4 @@
+gmailAccessToken = ""
+openAIKey = ""
+salesforceBaseUrl = "https://instance-id.salesforce.com"
+salesforceAccessToken = ""
```
Comment (anupama-pathirage) [inline salesforce_api/gmail-to-salesforce-lead/Config.toml.example]: Avoid committing config.toml

Comment (kaneeldias) [reply salesforce_api/gmail-to-salesforce-lead/Config.toml.example]: This is an example Config.toml file for the user to get an idea of the structure. The actual Config.toml file is not committed.

File: salesforce_api/gmail-to-salesforce-lead/main.bal (hunk 1 above)
Comment (kaneeldias) [reply salesforce_api/gmail-to-salesforce-lead/main.bal]: Fixed in [2ca5fbc](https://github.com/ballerina-guides/integration-samples/pull/32/commits/2ca5fbcc2db32d6c70c870bc842b0fd10662a08b)

File: salesforce_api/gmail-to-salesforce-lead/Config.toml.example (hunk 2 above)
Comment (anupama-pathirage) [reply salesforce_api/gmail-to-salesforce-lead/Config.toml.example]: For other samples in this repo we followed the practice of adding config.toml structure to readme. So let's follow the same for all samples

File: gmail-to-salesforce-lead/README.md [hunk 3]
```diff
@@ -0,0 +1,80 @@
+# Example to create leads obtained via email on Salesforce using the OpenAI API.
```
Comment (anupama-pathirage) [inline gmail-to-salesforce-lead/README.md]: Follow the pattern in existing samples. It should be something like
`Email lead info into Salesforce using OpenAI` Refer to https://github.com/ballerina-guides/integration-samples/blob/main/gcalendar-new-event-to-trello-card-and-twillio-msg/README.md etc

File: gmail-to-salesforce-lead/README.md [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+## Use case
+The following sample demonstrates a scenario in which customer leads obtained through email are automatically pushed to Salesforce. The required details for the lead (name, company, designation etc.) are inferred using the content of the email and the OpenAI chat API.
```
Comment (anupama-pathirage) [inline gmail-to-salesforce-lead/README.md]:
```suggestion
The following sample demonstrates a scenario in which customer leads obtained through email are automatically pushed to Salesforce. The required details for the lead (name, company, designation, etc.) are inferred using the content of the email and the OpenAI chat API.
```

File: gmail-to-salesforce-lead/README.md [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+When the user receives an email pertaining to a lead, they will mark that thread with a specific label (e.g. `"Lead"`). This Ballerina program will run continuously in the background, polling the email server every 10 minutes for threads marked with this label. If an email is found, its content will be read and used to infer the following details
```
Comment (anupama-pathirage) [inline gmail-to-salesforce-lead/README.md]:
```suggestion
When the user receives an email pertaining to a lead, they will mark that thread with a specific label (e.g., `"Lead"`). This Ballerina program will run continuously in the background, polling the email server every 10 minutes for threads marked with this label. If an email is found, its content will be read and used to infer the following details
```

File: gmail-to-salesforce-lead/README.md (hunk 3 above)
Comment (anupama-pathirage) [reply gmail-to-salesforce-lead/README.md]: Refer to the same link above and add single sentence with purpose. E.g. `This sample creates a card in trello board and send a Twillio SMS, for each new event created in Google Calendar.` 

There link to salesforce and OpenAI

File: gmail-to-salesforce-lead/README.md [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+Once these details have been inferred, a new lead will be generated on Salesforce.
+
+![Flow diagram](/docs/flow-diagram.png)
+
+## Prerequisites
+* An email account configured to use Gmail
+* An account on the Google Cloud Platform
+* An OpenAI account with API usage enabled
+* A Salesforce account
```
Comment (anupama-pathirage) [inline gmail-to-salesforce-lead/README.md]: Provide links to all above platforms.

File: gmail-to-salesforce-lead/README.md [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+
+>Note: The following steps will require you to generate keys for the Gmail, OpenAI and Salesforce APIs. These keys will have to be securely stored in the `Config.toml` file in the project directory under the relevant fields. To see a template of this file, please see the [`Config.toml.example`] file (./Config.toml.example).
```
Comment (anupama-pathirage) [inline gmail-to-salesforce-lead/README.md]:
```suggestion
>Note: The following steps will require you to generate keys for the Gmail, OpenAI, and Salesforce APIs. These keys will have to be securely stored in the `Config.toml` file in the project directory under the relevant fields. 
```

Comment (anupama-pathirage) [reply gmail-to-salesforce-lead/README.md]: Instead give a sample config.toml file content within readme. Refer https://github.com/ballerina-guides/integration-samples/blob/main/gcalendar-new-event-to-trello-card-and-twillio-msg/README.md#configuration

File: gmail-to-salesforce-lead/README.md [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+### Configuring your email account to use Gmail
+> Note: If you already have a Gmail account (ending with `@gmail.com`) or your acccount is on the Google workspace, you do not need to follow the steps below. In essence, if you can access your email via `www.gmail.com`, the following is not necessary.
```
Comment (anupama-pathirage) [inline gmail-to-salesforce-lead/README.md]:
```suggestion
> Note: If you already have a Gmail account (ending with `@gmail.com`) or your account is on the Google workspace, you do not need to follow the steps below. In essence, if you can access your email via `www.gmail.com`, the following is not necessary.
```

File: gmail-to-salesforce-lead/README.md [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+1. Visit [Gmail](https://gmail.com) and create a new account or log into an existing account.
+2. Enter the `Accounts` tab under settings and click on `Add a mail account`.
+3. Provide the necessary authentication details to your email account.
+4. After adding a mail account, you should be able to see all new emails received to your email via the Gmail interface.
+
+### Obtaining the Gmail API keys
+1. Create a new [Google Cloud Platform project](https://console.cloud.google.com). 
+2. Find and click `APIs & Services` --> `Library` from the navigation menu.
+3. In the searchbox enter `"Gmail"`.
```
Comment (anupama-pathirage) [inline gmail-to-salesforce-lead/README.md]:
```suggestion
3. In the searchbox, enter `"Gmail"`.
```

File: gmail-to-salesforce-lead/README.md [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+4. Then select Gmail API and click `Enable` button.
+5. Complete OAuth consent screen setup.
```
Comment (anupama-pathirage) [inline gmail-to-salesforce-lead/README.md]:
```suggestion
5. Complete the OAuth consent screen setup.
```

File: gmail-to-salesforce-lead/README.md [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+6. Click the `Credential` tab from the left sidebar. In the displaying window click on the `Create Credentials` button Select OAuth client ID.
```
Comment (anupama-pathirage) [inline gmail-to-salesforce-lead/README.md]:
```suggestion
6. Click the `Credential` tab from the left sidebar. In the displaying window, click on the `Create Credentials` button and select OAuth client ID.
```

File: gmail-to-salesforce-lead/README.md [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+7. Fill in the required fields. Add `"https://developers.google.com/oauthplayground"` to the Redirect URI field.
+8. Note down the `clientId` and `clientSecret`.
+9. Visit https://developers.google.com/oauthplayground/. Go to settings (Top right corner) -> Tick 'Use your own OAuth credentials' and insert Oauth ClientId and clientSecret. Click close.
+10. Then, Complete Step1 (Select and Authorize APIs)
+11. Make sure you select the `"https://www.googleapis/auth/gmail.modify"` and `"https://www.googleapis/auth/gmail.labels` OAuth scopes. These two scopes will allow the program to read emails including adding/removing labels.
```
Comment (anupama-pathirage) [inline gmail-to-salesforce-lead/README.md]:
```suggestion
11. Make sure you select the `"https://www.googleapis/auth/gmail.modify"` and `"https://www.googleapis/auth/gmail.labels` OAuth scopes. These two scopes will allow the program to read emails, including adding/removing labels.
```

File: gmail-to-salesforce-lead/README.md [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+12. Click `Authorize APIs` and You will be in Step 2.
```
Comment (anupama-pathirage) [inline gmail-to-salesforce-lead/README.md]:
```suggestion
12. Click `Authorize APIs`, and You will be in Step 2.
```

File: gmail-to-salesforce-lead/main.bal [hunk 14]
```diff
@@ -0,0 +1,193 @@
...
+configurable string openAIKey = ?;
+configurable string salesforceBaseUrl = ?;
+configurable string salesforceAccessToken = ?;
+
+final string label = "Lead";
+
+public function main() returns error? {
+    while true {
+        Email[] emails = check getEmails(label);
+
```
Comment (anupama-pathirage) [inline gmail-to-salesforce-lead/main.bal]: Avoid these unnecessary new lines within the code. Check all the places.

File: gmail-to-salesforce-lead/main.bal [hunk 15]
```diff
@@ -0,0 +1,193 @@
...
+        };
+
+        check addLeadsToSalesforce(leads);
+
+        runtime:sleep(600);
+    }
+}
+
+function getEmails(string label) returns Email[]|error {
+    gmail:Client gmailClient = check new ({auth: {token: gmailAccessToken}});
```
Comment (anupama-pathirage) [inline gmail-to-salesforce-lead/main.bal]: It seems email client is created during each iteration of the loop. instead can we defined a final client outside the main?

File: gmail-to-salesforce-lead/main.bal [hunk 16]
```diff
@@ -0,0 +1,193 @@
...
+
+    string[] labelIdsToMatch = check getLabelIds(gmailClient, [label]);
+    if (labelIdsToMatch.length() == 0) {
+        error e = error("Unable to find any labels to match.");
```
Comment (anupama-pathirage) [inline gmail-to-salesforce-lead/main.bal]:
```suggestion
        return error("Unable to find any labels to match.");
```

File: gmail-to-salesforce-lead/main.bal [hunk 17]
```diff
@@ -0,0 +1,193 @@
...
+function parseEmail(gmail:Message message) returns Email|error {
+    string 'from = <string>message.headerFrom;
+    string subject = <string>message.headerSubject;
+    string body = <string>(check mime:base64Decode(<string>(<gmail:MessageBodyPart>message.emailBodyInText).data));
+
+    return {
+        'from: 'from,
+        subject: subject,
+        body: body
+    };
```
Comment (anupama-pathirage) [inline gmail-to-salesforce-lead/main.bal]: Can simplify the method as follows. 

```ballerina
return {
        'from: <string>message.headerFrom,
        subject: <string>message.headerFrom,
        body: <string>(check mime:base64Decode(<string>(<gmail:MessageBodyPart>message.emailBodyInText).data))
    };
```

File: gmail-to-salesforce-lead/main.bal [hunk 18]
```diff
@@ -0,0 +1,193 @@
...
+}
+
+function generateLead(string 'from, string subject, string body) returns Lead|error {
+    openAI:Client openAIClient = check new ({
```
Comment (anupama-pathirage) [inline gmail-to-salesforce-lead/main.bal]: Seems this client is also created during each iteration of loop.

File: gmail-to-salesforce-lead/main.bal [hunk 19]
```diff
@@ -0,0 +1,193 @@
...
+                role: "user",
+                content: string `
+            Extract the following details in JSON from the email.
+                {
+                    firstName__c: string, // Mandatory
+                    lastName__c: string, // Mandatory
+                    email__c: string // Mandatory
+                    phoneNumber__c: string, // With country code. Use N/A if unable to find
+                    company__c: string, // Mandatory
+                    designation__c: string // Not mandator. Use N/A if unable to find
```
Comment (anupama-pathirage) [inline gmail-to-salesforce-lead/main.bal]:
```suggestion
                    designation__c: string // Not mandatory. Use N/A if unable to find
```

File: gmail-to-salesforce-lead/main.bal [hunk 20]
```diff
@@ -0,0 +1,193 @@
...
+                body: ${body}
+            }
+        `
+            }
+        ]
+    };
+
+    openAI:CreateChatCompletionResponse response = check openAIClient->/chat/completions.post(request);
+
+    Lead result = check (<string>response.choices[0].message?.content).fromJsonStringWithType(Lead);
```
Comment (anupama-pathirage) [inline gmail-to-salesforce-lead/main.bal]:
```suggestion
    return check (<string>response.choices[0].message?.content).fromJsonStringWithType(Lead);
```

File: gmail-to-salesforce-lead/main.bal [hunk 21]
```diff
@@ -0,0 +1,193 @@
...
+    return result;
```
Comment (anupama-pathirage) [inline gmail-to-salesforce-lead/main.bal]:
```suggestion
```

File: gmail-to-salesforce-lead/main.bal [hunk 22]
```diff
@@ -0,0 +1,193 @@
...
+}
+
+function addLeadsToSalesforce(Lead[] leads) returns error? {
+    sfdc:Client sfdcClient = check new ({
```
Comment (anupama-pathirage) [inline gmail-to-salesforce-lead/main.bal]: This client is also created during each iteration of the loop

File: gmail-to-salesforce-lead/main.bal [hunk 23]
```diff
@@ -0,0 +1,193 @@
...
+        baseUrl: salesforceBaseUrl,
+        auth: {token: salesforceAccessToken}
+    });
+
+    from Lead lead in leads
+    do {
+        sfdc:CreationResponse|error createResponse = check sfdcClient->create("EmailLead__c", lead);
+        if createResponse is error {
+            log:printError("An error occured while creating a Lead object on salesforce.", createResponse, createResponse.stackTrace(), lead = lead);
```
Comment (anupama-pathirage) [inline gmail-to-salesforce-lead/main.bal]: Break at max char length of 120 https://learn-ballerina.github.io/best_practices/wrap_code.html

File: gmail-to-salesforce-lead/main.bal [hunk 24]
```diff
@@ -0,0 +1,193 @@
...
+
+function addLeadsToSalesforce(Lead[] leads) returns error? {
+    sfdc:Client sfdcClient = check new ({
+        baseUrl: salesforceBaseUrl,
+        auth: {token: salesforceAccessToken}
+    });
+
+    from Lead lead in leads
+    do {
+        sfdc:CreationResponse|error createResponse = check sfdcClient->create("EmailLead__c", lead);
```
Comment (anupama-pathirage) [inline gmail-to-salesforce-lead/main.bal]: How the error is returned as `check` is used in this line?

File: gmail-to-salesforce-lead/README.md [hunk 25]
```diff
@@ -0,0 +1,80 @@
...
+* First name
+* Last name
+* Phone number
+* Email address
+* Company name
+* Designation
+
+Once these details have been inferred, a new lead will be generated on Salesforce.
+
+![Flow diagram](/docs/flow-diagram.png)
```
Comment (MaryamZi) [inline gmail-to-salesforce-lead/README.md]: Link is broken?

File: gmail-to-salesforce-lead/README.md (hunk 4 above)
Comment (MaryamZi) [reply gmail-to-salesforce-lead/README.md]: "inferred from" instead of "inferred using"?

File: gmail-to-salesforce-lead/docs/images/flow-diagram.png
Comment (MaryamZi) [inline gmail-to-salesforce-lead/docs/images/flow-diagram.png:1]: e-mail vs emails inconsistent.

File: gmail-to-salesforce-lead/main.bal [hunk 26]
```diff
@@ -0,0 +1,193 @@
+//Re-order the imports alphabetically
+import ballerina/lang.runtime;
+import ballerina/log;
+import ballerina/mime;
+import ballerinax/googleapis.gmail as gmail;
```
Comment (MaryamZi) [inline gmail-to-salesforce-lead/main.bal]:
```suggestion
import ballerinax/googleapis.gmail;
```

Defaults to `gmail`.

File: gmail-to-salesforce-lead/main.bal [hunk 27]
```diff
@@ -0,0 +1,193 @@
...
+    string company__c;
+    string designation__c;
+|};
+
+configurable string gmailAccessToken = ?;
+configurable string openAIKey = ?;
+configurable string salesforceBaseUrl = ?;
+configurable string salesforceAccessToken = ?;
+
+final string label = "Lead";
```
Comment (MaryamZi) [inline gmail-to-salesforce-lead/main.bal]: Why not a const?

File: gmail-to-salesforce-lead/main.bal [hunk 28]
```diff
@@ -0,0 +1,193 @@
...
+        Email[] emails = check getEmails(label);
+
+        Lead[] leads = [];
+        from Email email in emails
+        do {
+            Lead|error lead = generateLead(email.'from, email.subject, email.body);
+            if lead is Lead {
+                leads.push(lead);
+            }
+        };
```
Comment (MaryamZi) [inline gmail-to-salesforce-lead/main.bal]:
```suggestion
        Lead[] leads = from Email email in emails 
                        let Lead|error lead = generateLead(email.'from, email.subject, email.body)
                        where lead is Lead
                        select lead;
```

File: gmail-to-salesforce-lead/main.bal [hunk 29]
```diff
@@ -0,0 +1,193 @@
...
+final string label = "Lead";
+
+public function main() returns error? {
+    while true {
+        Email[] emails = check getEmails(label);
+
+        Lead[] leads = [];
+        from Email email in emails
+        do {
+            Lead|error lead = generateLead(email.'from, email.subject, email.body);
```
Comment (MaryamZi) [inline gmail-to-salesforce-lead/main.bal]: I would pass email instead and extract the fields out within the function.

File: gmail-to-salesforce-lead/main.bal [hunk 30]
```diff
@@ -0,0 +1,193 @@
...
+
+        runtime:sleep(600);
+    }
+}
+
+function getEmails(string label) returns Email[]|error {
+    gmail:Client gmailClient = check new ({auth: {token: gmailAccessToken}});
+
+    string[] labelIdsToMatch = check getLabelIds(gmailClient, [label]);
+    if (labelIdsToMatch.length() == 0) {
```
Comment (MaryamZi) [inline gmail-to-salesforce-lead/main.bal]: Parentheses are not required.

File: gmail-to-salesforce-lead/main.bal [hunk 31]
```diff
@@ -0,0 +1,193 @@
...
+
+    gmail:MailThread[] matchingMailThreads = check getMatchingMailThreads(gmailClient, labelIdsToMatch);
+    removeLabels(gmailClient, matchingMailThreads, labelIdsToMatch);
+
+    gmail:Message[] matchingEmails = getMatchingEmails(gmailClient, matchingMailThreads);
+
+    Email[] emails = [];
+    from gmail:Message message in matchingEmails
+    do {
+        Email|error email = parseEmail(message);
```
Comment (MaryamZi) [inline gmail-to-salesforce-lead/main.bal]: Can use a query expression here also. Check wherever applicable.

File: gmail-to-salesforce-lead/main.bal [hunk 32]
```diff
@@ -0,0 +1,193 @@
...
+    
+    return emails;
+}
+
+function getLabelIds(gmail:Client gmailClient, string[] labelsToMatch) returns string[]|error {
+    gmail:LabelList labelList = check gmailClient->listLabels("me");
+
+    return from gmail:Label label in labelList.labels
+        where labelsToMatch.indexOf(label.name) != ()
+        select label.id;
```
Comment (MaryamZi) [inline gmail-to-salesforce-lead/main.bal]:
```suggestion
    return from gmail:Label {name, id} in labelList.labels
        where labelsToMatch.indexOf(name) != ()
        select id;
```

File: gmail-to-salesforce-lead/main.bal [hunk 33]
```diff
@@ -0,0 +1,193 @@
...
+
+function getMatchingMailThreads(gmail:Client gmailClient, string[] labelIdsToMatch) returns gmail:MailThread[]|error {
+    gmail:MsgSearchFilter searchFilter = {
+        includeSpamTrash: false,
+        labelIds: labelIdsToMatch
+    };
+
+    stream<gmail:MailThread, error?>|error mailThreadStream = gmailClient->listThreads(filter = searchFilter);
+    if mailThreadStream is error {
+        return mailThreadStream;
```
Comment (MaryamZi) [inline gmail-to-salesforce-lead/main.bal]: Why aren't you using `check` instead?

File: gmail-to-salesforce-lead/main.bal [hunk 34]
```diff
@@ -0,0 +1,193 @@
...
+    }
+
+    return check from gmail:MailThread mailThread in mailThreadStream
```
Comment (MaryamZi) [inline gmail-to-salesforce-lead/main.bal]: `check` is unnecessary here, just results in an additional `is` check.

File: gmail-to-salesforce-lead/main.bal [hunk 35]
```diff
@@ -0,0 +1,193 @@
...
+        select mailThread;
+}
+
+function removeLabels(gmail:Client gmailClient, gmail:MailThread[] mailThreads, string[] labelIds) {
+    from gmail:MailThread mailThread in mailThreads
+    do {
```
Comment (MaryamZi) [inline gmail-to-salesforce-lead/main.bal]: Why do you need a query action? Why not a simple for loop?

File: gmail-to-salesforce-lead/main.bal [hunk 36]
```diff
@@ -0,0 +1,193 @@
...
+        gmail:MailThread|error removeLabelResponse = gmailClient->modifyThread(mailThread.id, [], labelIds);
+        if removeLabelResponse is error {
+            log:printError("An error occured in removing the labels from the thread.", removeLabelResponse, removeLabelResponse.stackTrace(), threadId = mailThread.id, labelIds = labelIds);
+        }
+    };
+}
+
+function getMatchingEmails(gmail:Client gmailClient, gmail:MailThread[] mailThreads) returns gmail:Message[] {
+    gmail:Message[] messages = [];
+    _ = from gmail:MailThread mailThread in mailThreads
```
Comment (MaryamZi) [inline gmail-to-salesforce-lead/main.bal]: `_ = ` is not necessary?

File: gmail-to-salesforce-lead/main.bal [hunk 37]
```diff
@@ -0,0 +1,193 @@
...
+        do {
+            gmail:MailThread|error response = gmailClient->readThread(mailThread.id);
+            if response is error {
+                log:printError("An error occured while reading the email.", response, response.stackTrace(), threadId = mailThread.id);
+            } else {
+                messages.push((<gmail:Message[]>response.messages)[0]);
```
Comment (MaryamZi) [inline gmail-to-salesforce-lead/main.bal]: Casts panic if the value don't belong to the target type. ensureType or an explicit is check will be better.

File: gmail-to-salesforce-lead/main.bal [hunk 38]
```diff
@@ -0,0 +1,193 @@
...
+            }
+        };
+
+    return messages;
+}
+
+function parseEmail(gmail:Message message) returns Email|error {
+    string 'from = <string>message.headerFrom;
+    string subject = <string>message.headerSubject;
+    string body = <string>(check mime:base64Decode(<string>(<gmail:MessageBodyPart>message.emailBodyInText).data));
```
Comment (MaryamZi) [inline gmail-to-salesforce-lead/main.bal]: Same comment re: casts.

File: salesforce_api/gmail-to-salesforce-lead/Config.toml.example (hunk 2 above)
Comment (kaneeldias) [reply salesforce_api/gmail-to-salesforce-lead/Config.toml.example]: Fixed in [f0cbe3a](https://github.com/ballerina-guides/integration-samples/pull/32/commits/f0cbe3a0dc7b75d145d3c3c8ba72893b86f18e2e)

File: gmail-to-salesforce-lead/README.md (hunk 4 above)
Comment (kaneeldias) [reply gmail-to-salesforce-lead/README.md]: Fixed in [a718fba](https://github.com/ballerina-guides/integration-samples/pull/32/commits/a718fba876d7ed04e212bd0cd384c113178380cb)

File: gmail-to-salesforce-lead/README.md (hunk 5 above)
Comment (kaneeldias) [reply gmail-to-salesforce-lead/README.md]: Fixed in [a718fba](https://github.com/ballerina-guides/integration-samples/pull/32/commits/a718fba876d7ed04e212bd0cd384c113178380cb)

File: gmail-to-salesforce-lead/README.md (hunk 6 above)
Comment (kaneeldias) [reply gmail-to-salesforce-lead/README.md]: Fixed in [a718fba](https://github.com/ballerina-guides/integration-samples/pull/32/commits/a718fba876d7ed04e212bd0cd384c113178380cb)

File: gmail-to-salesforce-lead/README.md (hunk 7 above)
Comment (kaneeldias) [reply gmail-to-salesforce-lead/README.md]: Fixed in [a718fba](https://github.com/ballerina-guides/integration-samples/pull/32/commits/a718fba876d7ed04e212bd0cd384c113178380cb)

File: gmail-to-salesforce-lead/README.md (hunk 3 above)
Comment (kaneeldias) [reply gmail-to-salesforce-lead/README.md]: Fixed in [a718fba](https://github.com/ballerina-guides/integration-samples/pull/32/commits/a718fba876d7ed04e212bd0cd384c113178380cb)

File: gmail-to-salesforce-lead/README.md (hunk 8 above)
Comment (kaneeldias) [reply gmail-to-salesforce-lead/README.md]: Fixed in [a718fba](https://github.com/ballerina-guides/integration-samples/pull/32/commits/a718fba876d7ed04e212bd0cd384c113178380cb)

File: gmail-to-salesforce-lead/README.md (hunk 9 above)
Comment (kaneeldias) [reply gmail-to-salesforce-lead/README.md]: Fixed in [a718fba](https://github.com/ballerina-guides/integration-samples/pull/32/commits/a718fba876d7ed04e212bd0cd384c113178380cb)

File: gmail-to-salesforce-lead/README.md (hunk 11 above)
Comment (kaneeldias) [reply gmail-to-salesforce-lead/README.md]: Fixed in [a718fba](https://github.com/ballerina-guides/integration-samples/pull/32/commits/a718fba876d7ed04e212bd0cd384c113178380cb)

File: gmail-to-salesforce-lead/README.md (hunk 13 above)
Comment (kaneeldias) [reply gmail-to-salesforce-lead/README.md]: Fixed in [a718fba](https://github.com/ballerina-guides/integration-samples/pull/32/commits/a718fba876d7ed04e212bd0cd384c113178380cb)

File: gmail-to-salesforce-lead/README.md (hunk 12 above)
Comment (kaneeldias) [reply gmail-to-salesforce-lead/README.md]: Fixed in [a718fba](https://github.com/ballerina-guides/integration-samples/pull/32/commits/a718fba876d7ed04e212bd0cd384c113178380cb)

File: gmail-to-salesforce-lead/README.md (hunk 25 above)
Comment (kaneeldias) [reply gmail-to-salesforce-lead/README.md]: Fixed in [a718fba](https://github.com/ballerina-guides/integration-samples/pull/32/commits/a718fba876d7ed04e212bd0cd384c113178380cb)

File: gmail-to-salesforce-lead/docs/images/flow-diagram.png
Comment (kaneeldias) [reply gmail-to-salesforce-lead/docs/images/flow-diagram.png:1]: Fixed in [a718fba](https://github.com/ballerina-guides/integration-samples/pull/32/commits/a718fba876d7ed04e212bd0cd384c113178380cb)

File: gmail-to-salesforce-lead/main.bal (hunk 14 above)
Comment (kaneeldias) [reply gmail-to-salesforce-lead/main.bal]: [cce2eaf](https://github.com/ballerina-guides/integration-samples/pull/32/commits/cce2eafff5580b9fc4de69c431a8951e6c90b36a)

File: gmail-to-salesforce-lead/main.bal (hunk 15 above)
Comment (kaneeldias) [reply gmail-to-salesforce-lead/main.bal]: Fixed in [cce2eaf](https://github.com/ballerina-guides/integration-samples/pull/32/commits/cce2eafff5580b9fc4de69c431a8951e6c90b36a)

File: gmail-to-salesforce-lead/main.bal (hunk 16 above)
Comment (kaneeldias) [reply gmail-to-salesforce-lead/main.bal]: Fixed in [cce2eaf](https://github.com/ballerina-guides/integration-samples/pull/32/commits/cce2eafff5580b9fc4de69c431a8951e6c90b36a)

File: gmail-to-salesforce-lead/main.bal (hunk 17 above)
Comment (kaneeldias) [reply gmail-to-salesforce-lead/main.bal]: Fixed in [cce2eaf](https://github.com/ballerina-guides/integration-samples/pull/32/commits/cce2eafff5580b9fc4de69c431a8951e6c90b36a)

File: gmail-to-salesforce-lead/main.bal (hunk 18 above)
Comment (kaneeldias) [reply gmail-to-salesforce-lead/main.bal]: Fixed in [cce2eaf](https://github.com/ballerina-guides/integration-samples/pull/32/commits/cce2eafff5580b9fc4de69c431a8951e6c90b36a)

File: gmail-to-salesforce-lead/main.bal (hunk 19 above)
Comment (kaneeldias) [reply gmail-to-salesforce-lead/main.bal]: Fixed in [cce2eaf](https://github.com/ballerina-guides/integration-samples/pull/32/commits/cce2eafff5580b9fc4de69c431a8951e6c90b36a)

File: gmail-to-salesforce-lead/main.bal (hunk 20 above)
Comment (kaneeldias) [reply gmail-to-salesforce-lead/main.bal]: Fixed in [cce2eaf](https://github.com/ballerina-guides/integration-samples/pull/32/commits/cce2eafff5580b9fc4de69c431a8951e6c90b36a)

File: gmail-to-salesforce-lead/main.bal (hunk 22 above)
Comment (kaneeldias) [reply gmail-to-salesforce-lead/main.bal]: Fixed in [cce2eaf](https://github.com/ballerina-guides/integration-samples/pull/32/commits/cce2eafff5580b9fc4de69c431a8951e6c90b36a)

File: gmail-to-salesforce-lead/main.bal (hunk 21 above)
Comment (kaneeldias) [reply gmail-to-salesforce-lead/main.bal]: Fixed in [cce2eaf](https://github.com/ballerina-guides/integration-samples/pull/32/commits/cce2eafff5580b9fc4de69c431a8951e6c90b36a)

File: gmail-to-salesforce-lead/main.bal (hunk 24 above)
Comment (kaneeldias) [reply gmail-to-salesforce-lead/main.bal]: Fixed in [cce2eaf](https://github.com/ballerina-guides/integration-samples/pull/32/commits/cce2eafff5580b9fc4de69c431a8951e6c90b36a)

File: gmail-to-salesforce-lead/main.bal (hunk 23 above)
Comment (kaneeldias) [reply gmail-to-salesforce-lead/main.bal]: Fixed in [cce2eaf](https://github.com/ballerina-guides/integration-samples/pull/32/commits/cce2eafff5580b9fc4de69c431a8951e6c90b36a)

File: gmail-to-salesforce-lead/main.bal (hunk 26 above)
Comment (kaneeldias) [reply gmail-to-salesforce-lead/main.bal]: Fixed in [cce2eaf](https://github.com/ballerina-guides/integration-samples/pull/32/commits/cce2eafff5580b9fc4de69c431a8951e6c90b36a)

File: gmail-to-salesforce-lead/main.bal (hunk 27 above)
Comment (kaneeldias) [reply gmail-to-salesforce-lead/main.bal]: Fixed in [cce2eaf](https://github.com/ballerina-guides/integration-samples/pull/32/commits/cce2eafff5580b9fc4de69c431a8951e6c90b36a)

File: gmail-to-salesforce-lead/main.bal (hunk 28 above)
Comment (kaneeldias) [reply gmail-to-salesforce-lead/main.bal]: Fixed in [cce2eaf](https://github.com/ballerina-guides/integration-samples/pull/32/commits/cce2eafff5580b9fc4de69c431a8951e6c90b36a)

File: gmail-to-salesforce-lead/main.bal (hunk 29 above)
Comment (kaneeldias) [reply gmail-to-salesforce-lead/main.bal]: Fixed in [cce2eaf](https://github.com/ballerina-guides/integration-samples/pull/32/commits/cce2eafff5580b9fc4de69c431a8951e6c90b36a)

File: gmail-to-salesforce-lead/main.bal (hunk 30 above)
Comment (kaneeldias) [reply gmail-to-salesforce-lead/main.bal]: Fixed in [cce2eaf](https://github.com/ballerina-guides/integration-samples/pull/32/commits/cce2eafff5580b9fc4de69c431a8951e6c90b36a)

File: gmail-to-salesforce-lead/main.bal (hunk 31 above)
Comment (kaneeldias) [reply gmail-to-salesforce-lead/main.bal]: Fixed in [cce2eaf](https://github.com/ballerina-guides/integration-samples/pull/32/commits/cce2eafff5580b9fc4de69c431a8951e6c90b36a)

File: gmail-to-salesforce-lead/main.bal (hunk 32 above)
Comment (kaneeldias) [reply gmail-to-salesforce-lead/main.bal]: Fixed in [cce2eaf](https://github.com/ballerina-guides/integration-samples/pull/32/commits/cce2eafff5580b9fc4de69c431a8951e6c90b36a)

File: gmail-to-salesforce-lead/main.bal (hunk 33 above)
Comment (kaneeldias) [reply gmail-to-salesforce-lead/main.bal]: Fixed in [cce2eaf](https://github.com/ballerina-guides/integration-samples/pull/32/commits/cce2eafff5580b9fc4de69c431a8951e6c90b36a)

File: gmail-to-salesforce-lead/main.bal (hunk 34 above)
Comment (kaneeldias) [reply gmail-to-salesforce-lead/main.bal]: Fixed in [cce2eaf](https://github.com/ballerina-guides/integration-samples/pull/32/commits/cce2eafff5580b9fc4de69c431a8951e6c90b36a)

File: gmail-to-salesforce-lead/main.bal (hunk 36 above)
Comment (kaneeldias) [reply gmail-to-salesforce-lead/main.bal]: Fixed in [cce2eaf](https://github.com/ballerina-guides/integration-samples/pull/32/commits/cce2eafff5580b9fc4de69c431a8951e6c90b36a)

File: gmail-to-salesforce-lead/main.bal (hunk 35 above)
Comment (kaneeldias) [reply gmail-to-salesforce-lead/main.bal]: Fixed in [cce2eaf](https://github.com/ballerina-guides/integration-samples/pull/32/commits/cce2eafff5580b9fc4de69c431a8951e6c90b36a)

File: gmail-to-salesforce-lead/main.bal (hunk 38 above)
Comment (kaneeldias) [reply gmail-to-salesforce-lead/main.bal]: Fixed in [cce2eaf](https://github.com/ballerina-guides/integration-samples/pull/32/commits/cce2eafff5580b9fc4de69c431a8951e6c90b36a)

File: gmail-to-salesforce-lead/main.bal (hunk 37 above)
Comment (kaneeldias) [reply gmail-to-salesforce-lead/main.bal]: Fixed in [cce2eaf](https://github.com/ballerina-guides/integration-samples/pull/32/commits/cce2eafff5580b9fc4de69c431a8951e6c90b36a)

**Review by anupama-pathirage — APPROVED** [review]

**Pull request merged**